File: backend_python/app/api/v1/sessions.py
"""Session transcript API endpoints."""
from typing import Optional
from fastapi import APIRouter, HTTPException, Query, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.services.session_service import SessionService
from app.models.schemas import (
    SessionListResponse,
    SessionProjectListResponse,
    SessionDetailResponse,
    SessionStatsResponse,
)

router = APIRouter()


@router.get("/sessions/projects", response_model=SessionProjectListResponse)
async def list_projects(db: AsyncSession = Depends(get_db)):
    """List all projects with session counts."""
    try:
        service = SessionService(db)
        return await service.list_projects()
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to list projects: {str(e)}")


@router.get("/sessions", response_model=SessionListResponse)
async def list_sessions(
    project_folder: Optional[str] = Query(None, description="Filter by project"),
    limit: int = Query(50, ge=1, le=100, description="Max sessions to return"),
    sort_by: str = Query("date", pattern="^(date|size)$"),
    sort_order: str = Query("desc", pattern="^(asc|desc)$"),
    db: AsyncSession = Depends(get_db),
):
    """List session summaries."""
    try:
        service = SessionService(db)
        return await service.list_sessions(project_folder, limit, sort_by, sort_order)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to list sessions: {str(e)}")


@router.get("/sessions/dashboard/stats", response_model=SessionStatsResponse)
async def get_dashboard_stats(db: AsyncSession = Depends(get_db)):
    """Get session statistics for dashboard."""
    try:
        service = SessionService(db)
        return await service.get_dashboard_stats()
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to get stats: {str(e)}")


@router.get("/sessions/{project_folder}/{session_id}", response_model=SessionDetailResponse)
async def get_session_detail(
    project_folder: str,
    session_id: str,
    page: int = Query(1, ge=1, description="Page number (5 prompts per page)"),
    db: AsyncSession = Depends(get_db),
):
    """Get full session detail with pagination."""
    try:
        service = SessionService(db)
        return await service.get_session_detail(session_id, project_folder, page)
    except FileNotFoundError:
        raise HTTPException(status_code=404, detail="Session not found")
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to get session: {str(e)}")

File: app/server/core/src/api/v1/sessions.rs
use crate::api::v1::ApiState;
use crate::services::session_service::{
    SessionDetailResponse, SessionListResponse, SessionProjectListResponse, SessionStatsResponse,
};
use axum::{
    Json, Router,
    extract::{Path, Query, State},
    http::StatusCode,
    routing::get,
};
use serde::Deserialize;

/// Default page size for the session list when the client doesn't pass `limit`.
const DEFAULT_LIST_LIMIT: usize = 100;

#[derive(Debug, Deserialize)]
struct DetailQuery {
    #[serde(default)]
    page: Option<usize>,
}

#[derive(Debug, Deserialize)]
struct ListQuery {
    #[serde(default)]
    project_folder: Option<String>,
    #[serde(default)]
    limit: Option<usize>,
    // sort_by / sort_order accepted for forward-compat; list is modified-desc.
    #[serde(default)]
    #[allow(dead_code)]
    sort_by: Option<String>,
    #[serde(default)]
    #[allow(dead_code)]
    sort_order: Option<String>,
}

pub fn router() -> Router<ApiState> {
    Router::new()
        .route("/", get(list_sessions))
        .route("/projects", get(get_session_projects))
        .route("/dashboard/stats", get(get_session_stats))
        .route("/{project}/{session}", get(get_session_detail))
}

async fn list_sessions(
    State(state): State<ApiState>,
    Query(query): Query<ListQuery>,
) -> Result<Json<SessionListResponse>, (StatusCode, String)> {
    let limit = query.limit.unwrap_or(DEFAULT_LIST_LIMIT).max(1);
    state
        .session_service
        .list_sessions(query.project_folder.as_deref(), limit)
        .await
        .map(Json)
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))
}

async fn get_session_projects(
    State(state): State<ApiState>,
) -> Result<Json<SessionProjectListResponse>, (StatusCode, String)> {
    state
        .session_service
        .list_projects()
        .await
        .map(Json)
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))
}

async fn get_session_stats(
    State(state): State<ApiState>,
) -> Result<Json<SessionStatsResponse>, (StatusCode, String)> {
    state
        .session_service
        .dashboard_stats()
        .await
        .map(Json)
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))
}

async fn get_session_detail(
    State(state): State<ApiState>,
    Path((project, session)): Path<(String, String)>,
    Query(query): Query<DetailQuery>,
) -> Result<Json<SessionDetailResponse>, (axum::http::StatusCode, String)> {
    let page = query.page.unwrap_or(1).max(1);

    let detail = state
        .session_service
        .get_session_detail(&project, &session, page)
        .await
        .map_err(|e| (axum::http::StatusCode::BAD_REQUEST, e.to_string()))?;

    Ok(Json(detail))
}
